minidot divided by the length of no sequence when no alignment passed its
filters, and by a weight of 0 for alignments of fewer than 7 matching
bases; a plot too tall for an int was converted to one anyway; and read
names went into PostScript strings unescaped, so a parenthesis in one
broke the file.

diff --git a/biochef_eps.h b/biochef_eps.h
new file mode 100644
index 0000000..e52ace5
--- /dev/null
+++ b/biochef_eps.h
@@ -0,0 +1,9 @@
+static void biochef_eps_Mstr(FILE *fp, double x, double y, const char *s)
+{
+	fprintf(fp, "%g %g (", (float)x, (float)y);
+	for (; *s; ++s) {
+		if (*s == '(' || *s == ')' || *s == '\\') fputc('\\', fp);
+		fputc(*s, fp);
+	}
+	fputs(") MS\n", fp);
+}
diff --git a/dotter.c b/dotter.c
index 66d0743..ab3ab39 100644
--- a/dotter.c
+++ b/dotter.c
@@ -7,6 +7,7 @@
 #include "sdict.h"
 #include "kvec.h"
 #include "eps.h"
+#include "biochef_eps.h"
 
 typedef struct {
 	uint32_t qn, qs, qe;
@@ -108,6 +109,10 @@ int biochef_upstream_main(int argc, char *argv[])
 		s->ml = r.ml;
 	}
 	paf_close(f);
+	if (h.n == 0) {
+		fprintf(stderr, "[E::%s] no alignment in %s is %d bases long on both sequences (-s) with %d matching bases (-m) and an identity of %g (-i)\n", __func__, argv[optind], min_span, min_match, min_iden);
+		return 1;
+	}
 
 	for (i = 0; i < 2; ++i) { // 0 for target; 1 for query
 		uint32_t n = d[i]->n_seq;
@@ -130,7 +135,7 @@ int biochef_upstream_main(int argc, char *argv[])
 				q->tot += (double)coor * w;
 				q->w += w;
 			}
-			for (j = 0; j < n; ++j) b[j].tot /= b[j].w;
+			for (j = 0; j < n; ++j) if (b[j].w) b[j].tot /= b[j].w;
 			ks_introsort_dty(n, b);
 		}
 		acclen[i] = (uint64_t*)calloc(n, 8);
@@ -138,6 +143,10 @@ int biochef_upstream_main(int argc, char *argv[])
 			acclen[i][a[i][j].i] = l, l += d[i]->seq[a[i][j].i].len;
 		totlen[i] = l;
 	}
+	if ((double)width / totlen[0] * totlen[1] + .499 > 1e7) {
+		fprintf(stderr, "[E::%s] the queries are %g times as long as the targets: a plot %d wide would be too tall\n", __func__, (double)totlen[1] / totlen[0], width);
+		return 1;
+	}
 	height = (int)((double)width / totlen[0] * totlen[1] + .499);
 	sx = (double)width / totlen[0];
 	sy = (double)height / totlen[1];
@@ -149,12 +158,12 @@ int biochef_upstream_main(int argc, char *argv[])
 	if (!no_label) {
 		// write x labels
 		for (i = 0; i < d[0]->n_seq; ++i)
-			eps_Mstr(stdout, (acclen[0][a[0][i].i] + .5 * d[0]->seq[a[0][i].i].len) * sx, font_size*.5, a[0][i].name);
+			biochef_eps_Mstr(stdout, (acclen[0][a[0][i].i] + .5 * d[0]->seq[a[0][i].i].len) * sx, font_size*.5, a[0][i].name);
 		eps_stroke(stdout);
 		fprintf(stdout, "gsave %g 0 translate 90 rotate\n", font_size*1.25);
 		// write y labels
 		for (i = 0; i < d[1]->n_seq; ++i)
-			eps_Mstr(stdout, (acclen[1][a[1][i].i] + .5 * d[1]->seq[a[1][i].i].len) * sx, 0, a[1][i].name);
+			biochef_eps_Mstr(stdout, (acclen[1][a[1][i].i] + .5 * d[1]->seq[a[1][i].i].len) * sx, 0, a[1][i].name);
 		fprintf(stdout, "grestore\n");
 		eps_stroke(stdout);
 	}
